cfg$cibersort			= '/home/huw/program/cibersort'
cfg$anaconda2 			= '/home/huw/program/anaconda2'
cfg$anaconda3 			= '/home/huw/program/anaconda3'
cfg$hsa.housekeeping.bed	= '/home/huw/program/hg19_RefSeq.bed'
cfg$hsaMethylBaitIntervalOrig='/home/huw/program/truseq-methyl-capture-epic-manifest-file.bed'
cfg$hsaMethylBaitInterval	='/home/huw/program/truseq-methyl-capture-epic-manifest-file-bait.interval'
cfg$hsaMethylTargetInterval	= '/home/huw/program/truseq-methyl-capture-epic-manifest-file-plusminus5bp.interval'
cfg$hsaMethylAnno		= '/home/huw/program/truseq-methyl-capture-epic-manifest-file.annotated.islands.txt'
cfg$hg19.cpgi			= '/home/huw/program/hg19_cpg_island.bed.txt'

cfg$fastqc 			= '/home/huw/program/FastQC/fastqc'
#options(scipen=999)
#system(paste0('cp /home/huw/program/truseq-methyl-capture-epic-header.txt ', hsaMethylBaitInterval))
#tmp = fread(hsaMethylBaitIntervalOrig)
#tmp = tmp[order(V1),]
#library(tibble)
#tmp = add_column(tmp, string = "+", .after = 3)
#tmp[, V1:=sub("chr", "", V1)]
#fwrite(tmp, file=hsaMethylBaitInterval, quote=F, sep="\t", col.names=F, append=T)
#write.table(as.data.frame(tmp), file=hsaMethylBaitInterval, quote=F, sep="\t", col.names=F, append=T, row.names=F)
#tmp[, V2:=V2-5]
#tmp[, V3:=V3+5]
#head(tmp)
#tmp[, V1 := as.character(V1)]
#tmp[, V2 := as.character(V2)]
#tmp[, V3 := as.character(V3)]
#tmp[, V4:=paste0("chr", V1, "_", V2, "-", V3)]
#system(paste0('cp /home/huw/program/truseq-methyl-capture-epic-header.txt ', hsaMethylTargetInterval))
#write.table(as.data.frame(tmp), file=hsaMethylTargetInterval, quote=F, sep="\t", col.names=F, append=T, row.names=F)
#rm(tmp)

cfg$bismarkGRCh37		= '/ifs/work/solitlab/huw/study/db/hsa/grch37_bismark'
cfg$bismark			= '/home/huw/program/Bismark'
cfg$methylTarget		= '/home/huw/program/truseq-methyl-capture-epic-manifest-file.bed'
cfg$SOAPHLA			= '/home/huw/program/SOAP-HLA'
cfg$bamReadCount		= '/home/huw/local/bam-readcount'
cfg$IEDBMHC1			= '/home/huw/program/mhc_i'
cfg$IEDBMHC1			= '/home/huw/program/mhc_ii'
cfg$pvacseq			= '/home/huw/program/anaconda3/bin/pvacseq'
cfg$tmpdir			= '/scratch/huw'
cfg$polysolver			= '/home/huw/program/polysolver'
cfg$ponfile			= '/home/huw/program/pon56Sample.vcf.gz'
cfg$make.trinuc			= '/home/huw/program/make_trinuc_maf.py'
cfg$mutsig			= '/opt/common/CentOS_6-dev/mutsig/cv-1.4/'
cfg$belly2			= '/home/huw/program/delly_v0.7.7_linux_x86_64bit'
cfg$exeDir			= '/ifs/work/solitlab/huw/program/socciCode/Pipelines/CBE/Variant/variants_pipeline/'
cfg$targetsDir			= '/ifs/work/solitlab/huw/program/socciCode/Pipelines/CBE/Variant/variants_pipeline/targets/'
cfg$facets			= '/ifs/work/solitlab/huw/program/socciCode/Pipelines/CBE/Variant/variants_pipeline/facets'
cfg$vepDir			= '/ifs/work/solitlab/huw/program/socciCode/Pipelines/CBE/Variant/variants_pipeline/vep'
cfg$dataDir			=  '/ifs/work/solitlab/huw/program/socciCode/Pipelines/CBE/Variant/variants_pipeline/data'
cfg$ConvertQualityScore    	= '/home/huw/program/ConvertQualityScore'
cfg$STAR			= '/home/huw/program/STAR-STAR_2.5.0a/bin/Linux_x86_64_static/STAR'
cfg$STAR_DIR			= '/home/huw/program/STAR-STAR_2.5.0a/bin/Linux_x86_64_static'
cfg$RSEM			= '/opt/common/CentOS_6/rsem/RSEM-1.2.25/'
cfg$KALLISTO			= '/home/huw/local/bin/kallisto '
##cfg$STAR			= '/opt/common/CentOS_6/star/STAR-STAR_2.5.0a/bin/Linux_x86_64/STAR'
cfg$samtools			= '/opt/common/CentOS_6-dev/bin/current/samtools'
cfg$samtoolsdir 		= '/opt/common/CentOS_6-dev/samtools/samtools-1.3.1'
cfg$sambamba			= '/home/huw/program/sambamba' 
cfg$java			= '/home/huw/program/jdk7/bin/java'
cfg$java			= '/opt/common/CentOS_6-dev/java/jdk1.8.0_31/bin/java'
cfg$picard			= '/home/huw/program/picard.jar'
cfg$htseqcountexe		= '/home/huw/local/bin/htseq-count'
cfg$NGSPLOTexe			= '/home/huw/program/ngsplot/bin/ngs.plot.r'
cfg$NGSFilter			= '/home/huw/program/ngs-filters/'
cfg$QSUB			= '/common/sge/bin/lx24-amd64/qsub'
cfg$BSUB			= '/common/lsf/9.1/linux2.6-glibc2.3-x86_64/bin/bsub'
cfg$trim_galore			= '/home/huw/program/trim_galore'
cfg$bowtie2			= '/home/huw/program/bowtie2/bowtie2'
cfg$bowtie2dir                  = '/home/huw/program/bowtie2'
cfg$macs14                      = '/home/huw/local/bin/macs14'
cfg$macs2                  	= '/home/huw/local/bin/macs2'
cfg$homerFindMotif              = '/home/huw/program/homer/bin/findMotifsGenome.pl'
cfg$homerAnnotatePeaks		= '/home/huw/program/homer/bin/annotatePeaks.pl'
cfg$tssCounts			= '/home/huw/program/tsscounts.pl'
cfg$bwa				= '/home/huw/program/bwa-0.7.12/bwa '
cfg$bedtools			= '/home/huw/program/bedtools/bin'
cfg$bam2bigwig			= '/home/huw/program/bam2bigwig.sh'
cfg$annotationFilter		= '/home/huw/program/annotationFilter.pl'

cfg$enhancerMm10Bed		= '/ifs/work/solitlab//huw/study/db/mmu/dna/Mus_musculus.GRCm38.79.enhancer.bed'
cfg$enhancerMm9Bed          	= '/home/program/PLtest/Enhancers_mm9_CreyghtonPNAS/PutativeEnhancers_mm9_5types_SortMerge300_noPromOL_GT50bp.txt'
cfg$tssMm9Bed              	= '/home/huw/program/PLtest/refFlat_mm9_TSSpm2kbSORTED.txt'
cfg$tssMm10Bed			= '/ifs/work/solitlab//huw/study/db/mmu/dna/Mus_musculus.GRCm38.79.chr.symbol.tss2kb.bed'
cfg$genomeMm9Bowtie2		= '/ifs/work/solitlab//huw/study/db/mmu/mm9-bowtie2/mm9'
cfg$genomeMm9StarDir		= '/ifs/work/solitlab//huw/study/db/mmu/mm9_star'
cfg$genomeMm9GTF		= '/ifs/work/solitlab//huw/study/db/mmu/dna/Mus_musculus.NCBIM37.67.chr.gtf'
cfg$genomeMm10Bowtie2		= '/ifs/work/solitlab//huw/study/db/mmu/mm10-bowtie2/mm10'
cfg$genomeMm10StarDir		= '/ifs/work/solitlab//huw/study/db/mmu/mm10star'
cfg$genomeMm10GTF		= '/ifs/work/solitlab//huw/study/db/mmu/dna/Mus_musculus.GRCm38.79.chr.gtf'
cfg$mm10chromsize		= '/ifs/work/solitlab//huw/study/db/mmu/dna/Mus_musculus.GRCm38.79.chrom.size'

cfg$ABRA			= '/opt/common/CentOS_6/abra2/abra2-2.07'
cfg$BCFTOOLS			= '/opt/common/CentOS_6/bcftools/bcftools-1.2/bin'
cfg$BEDTOOLS			= '/opt/common/CentOS_6/bedtools/bedtools-2.22.0/bin'
cfg$BWA				= '/opt/common/CentOS_6/bwa/bwa-0.7.12/bwa'
cfg$CUTADAPT			= '/opt/common/CentOS_6/cutadapt/cutadapt-1.9.1/bin/cutadapt'
cfg$DELLY			= '/opt/common/CentOS_6/delly/delly_v0.6.1'
cfg$dRANGER			= '/opt/common/CentOS_6/dranger/dRanger_annotate_v2.0'
cfg$FACETS_SUITE		= '/home/huw/program/facets-suite-1.0.1'
cfg$FACETS_LIB			= '/opt/common/CentOS_6-dev/facets_lib/facets-0.5.6'
cfg$FIXMULTIINDEL		= '/opt/common/CentOS_6/FixMultiInDel/FixMultiInDel-2.0.1'
cfg$GATK			= '/home/huw/program/GenomeAnalysisTK-3.8-0-ge9d806836'
cfg$GATK4			= '/home/huw/program/gatk-4.0.2.1/gatk'
#cfg$GATK4			= '/home/huw/program/gatk-4.0.2.1/gatk-package-4.0.2.1-local.jar'
cfg$gnomadfile			= '/home/huw/af-only-gnomad.raw.sites.b37.vcf.gz'


cfg$JAVA			= '/opt/common/CentOS_6/java/jdk1.8.0_31/bin'
cfg$JAVA7_MUTECT            	= '/opt/common/CentOS_6/java/jdk1.7.0_75/bin'
cfg$MCR				= '/opt/common/CentOS_6/matlab/v81'
cfg$MUTECT			= '/opt/common/CentOS_6/muTect/muTect-1.1.7'
cfg$PERL			= '/opt/common/CentOS_6/perl/perl-5.22.0/bin'
cfg$PERL			= '/home/huw/local/perl/perl-5.26.0/bin'
cfg$PICARD			= '/home/huw/program/picard.jar'
cfg$PYTHON			= '/opt/common/CentOS_6/python/python-2.7.8/bin/python'
cfg$R				= '/opt/common/CentOS_6/R/R-3.1.2/bin'
cfg$SAMTOOLS                	= '/opt/common/CentOS_6/samtools/samtools-1.2'
cfg$SCALPEL			= '/opt/common/CentOS_6/scalpel/scalpel-0.2.2'
cfg$SOMATICSNIPER		= '/opt/common/CentOS_6/somaticsniper/somatic-sniper-1.0.4'
cfg$STRELKA			= '/opt/common/CentOS_6/strelka/strelka_1.0.11'
cfg$TABIX			= '/opt/common/CentOS_6/samtools/samtools-1.2/htslib-1.2.1'
cfg$VARSCAN			= '/opt/common/CentOS_6/varscan/v2.3.7'
cfg$VCF2MAF			= '/home/huw/program/vcf2maf-1.6.14'
cfg$VCFTOOLS			= '/opt/common/CentOS_6-dev/vcftools/v0.1.14/bin'
cfg$VEP                     	= '/opt/common/CentOS_6-dev/vep/v86'
cfg$VEPv88                  	= '/opt/common/CentOS_6-dev/vep/v88/variant_effect_predictor.pl'
cfg$VEP_plugins             	= '/home/huw/program/VEP_plugins'
cfg$VIRMID			= '/opt/common/CentOS_6/virmid/Virmid-1.1.1'
cfg$WES_FILTER			= '/opt/common/CentOS_6/wes-filter/wes-filters-1.1.3'
cfg$CMOBIN			= '/opt/common/CentOS_6-dev/python/python-2.7.10/bin/'


cfg$B37_BWA_INDEX		= '/ifs/depot/assemblies/H.sapiens/b37/index/bwa/0.7.12/b37.fasta'
cfg$B37_FASTA			= '/ifs/depot/assemblies/H.sapiens/b37/b37.fasta'
cfg$B37_FAI			= '/ifs/depot/assemblies/H.sapiens/b37/b37.fasta.fai'
cfg$B37_MM10_HYBRID_BWA_INDEX	= '/ifs/depot/assemblies/hybrids/H.sapiens_M.musculus/b37_mm10/index/bwa/0.7.12/b37_mm10.fasta'
cfg$B37_MM10_HYBRID_FASTA	= '/ifs/depot/assemblies/hybrids/H.sapiens_M.musculus/b37_mm10/b37_mm10.fasta'
cfg$B37_MM10_HYBRID_FAI		= '/ifs/depot/assemblies/hybrids/H.sapiens_M.musculus/b37_mm10/b37_mm10.fasta.fai'
cfg$HG19_BWA_INDEX		= '/ifs/depot/assemblies/H.sapiens/hg19/index/bwa/0.7.4-r385/hg19.fasta'
cfg$HG19_FASTA			= '/ifs/depot/assemblies/H.sapiens/hg19/hg19.fasta'
cfg$HG19_FAI			= '/ifs/depot/assemblies/H.sapiens/hg19/hg19.fasta.fai'
cfg$HG19_MM10_HYBRID_BWA_INDEX	= '/ifs/depot/assemblies/hybrids/H.sapiens_M.musculus/hg19_mm10/index/bwa/0.7.8/hg19_mm10.fasta'
cfg$HG19_MM10_HYBRID_FASTA	= '/ifs/depot/assemblies/hybrids/H.sapiens_M.musculus/hg19_mm10/hg19_mm10.fasta'
cfg$HG19_MM10_HYBRID_FAI	= '/ifs/depot/assemblies/hybrids/H.sapiens_M.musculus/hg19_mm10/hg19_mm10.fasta.fai'
cfg$MM9_BWA_INDEX		= '/ifs/depot/assemblies/M.musculus/mm9/index/bwa/0.7.4-r385/mm9.fasta'
cfg$MM9_FASTA			= '/ifs/depot/assemblies/M.musculus/mm9/mm9.fasta'
cfg$MM9_FAI			= '/ifs/depot/assemblies/M.musculus/mm9/mm9.fasta.fai'
cfg$MM10_BWA_INDEX		= '/ifs/depot/assemblies/M.musculus/mm10/index/bwa/0.7.8/mm10.fasta'
cfg$MM10_FASTA			= '/ifs/depot/assemblies/M.musculus/mm10/mm10.fasta'
cfg$MM10_FAI			= '/ifs/depot/assemblies/M.musculus/mm10/mm10.fasta.fai'
cfg$clipR1			= 'AGATCGGAAGAGCACACGTCT'
cfg$clipR2			= 'AGATCGGAAGAGCACACGTCT'
cfg$bqtrim			= '3'

## for homer annotatePeaks.pl mm10
cfg$homerMm10AnnotationFile	= '/ifs/work/solitlab//huw/study/db/mmu/dna/Mus_musculus.GRCm38.79.chr.gtf.annotations.final.txt'

cfg$genomeHg19Bwa		= '/ifs/work/solitlab//huw/study/db/hsa/hg37-bwa/hg37 '
cfg$genomeHg19StarDir		= '/ifs/work/solitlab//huw/study/db/hsa/hg37star'
cfg$genomeHg19Bowtie2		= '/ifs/work/solitlab//huw/study/db/hsa/hg37bowtie2/hg37'
cfg$genomeHg38Bowtie2		= '/ifs/work/solitlab//huw/study/db/hsa/hg38_bowtie2/hg38'
cfg$genomeHg19GTF		= '/ifs/work/solitlab//huw/study/db/hsa/dna/Homo_sapiens.GRCh37.75.gtf'
cfg$hg19chromsize		= '/ifs/work/solitlab//huw/study/db/hsa/dna/Homo_sapiens.GRCh37.75.chrom.size'

cfg$genomeGRCh37Fasta		= '/ifs/work/solitlab/huw/study/db/hsa/dna/Homo_sapiens.GRCh37.75.dna.primary_assembly.fa'
cfg$genomeGRCh37StarDir		= '/ifs/depot/assemblies/H.sapiens/b37/index/star/2.4.1d/gencode/v18/overhang49 /ifs/work/solitlab/huw/park/study/hiseq/chipseq/template.conf'

cfg$genomeGRCh38GTF		= '/ifs/work/solitlab/huw/study/db/hsa/dna/Homo_sapiens.GRCh38.88.gtf'
cfg$genomeGRCh38Fasta		= '/ifs/work/solitlab/huw/study/db/hsa/dna/Homo_sapiens.GRCh38.dna.primary_assembly.fa'
cfg$genomeGRCh38StarDir		= '/ifs/work/solitlab/huw/study/db/hsa/hg38_star'
cfg$genomeGRCh38Rsem		= '/ifs/work/solitlab/huw/program/RSEM_tutorial/ref/GRCh38/GRCh38'
cfg$RSEM_REF			= '/ifs/work/solitlab/huw/program/RSEM_tutorial/ref/GRCh38/GRCh38'

cfg$genomeGRCh38Kallisto	= '/ifs/work/solitlab/huw/study/db/hsa/hg38_kallisto/kallisto_GRCh38'

cfg$genomeGRCm38Fasta		= '/ifs/work/solitlab/huw/study/db/mmu/dna/Mus_musculus.GRCm38.79.dna.primary_assembly.chr.fa'
cfg$genomeGRCm38GTF		= '/ifs/work/solitlab/huw/study/db/mmu/dna/Mus_musculus.GRCm38.79.chr.gtf'
cfg$genomeGRCm38Rsem		= '/ifs/work/solitlab/huw/program/RSEM_tutorial/ref/GRCm38/GRCm38'
